src/main/scala/vectorpipe/osm/internal/functions/package.scala

```diff
@@ -0,0 +1,636 @@
+    ),
+    "landuse" -> Map(),
+    "leisure" -> Map(
+      "slipway" -> true,
+      "track" -> true
+    ),
+    "man_made" -> Map(
+      "breakwater" -> true,
+      "crane" -> true,
+      "cutline" -> true,
+      "embankment" -> true,
+      "groyne" -> true,
+      "pier" -> true,
+      "pipeline" -> true
+    ),
+    "military" -> Map(),
+    "natural" -> Map(
+      "cliff" -> true,
+      "coastline" -> true,
+      "ridge" -> true,
+      "tree_row" -> true
+    ),
+    "office" -> Map(),
+    "piste:type" -> Map(),
+    "place" -> Map(),
+    "playground" -> Map(
+      "balancebeam" -> true,
+      "slide" -> true,
+      "zipwire" -> true
+    ),
+    "power" -> Map(
+      "line" -> true,
+      "minor_line" -> true
+    ),
+    "public_transport" -> Map(
+      "platform" -> true
+    ),
+    "shop" -> Map(),
+    "tourism" -> Map(),
+    "waterway" -> Map(
+      "canal" -> true,
+      "dam" -> true,
+      "ditch" -> true,
+      "drain" -> true,
+      "river" -> true,
+      "stream" -> true,
+      "weir" -> true
+    )
+  )
+
+  private val MultiPolygonTypes = Set("multipolygon", "boundary")
+
+  private val BooleanValues = Set("yes", "no", "true", "false", "1", "0")
+
+  private val TruthyValues = Set("yes", "true", "1")
+
+  private lazy val logger = Logger.getLogger(getClass)
+
+  private val _isArea = (tags: Map[String, String]) =>
+    tags match {
+      case _ if tags.contains("area") && BooleanValues.contains(tags("area").toLowerCase) =>
+        TruthyValues.contains(tags("area").toLowerCase)
+      case _ =>
+        // see https://github.com/osmlab/id-area-keys (values are inverted)
+        val matchingKeys = tags.keySet.intersect(AreaKeys.keySet)
+        matchingKeys.exists(k => !AreaKeys(k).contains(tags(k)))
+    }
+
+  val isArea: UserDefinedFunction = udf(_isArea)
+
+  private val _isMultiPolygon = (tags: Map[String, String]) =>
+    tags.contains("type") && MultiPolygonTypes.contains(tags("type").toLowerCase)
+
+  val isMultiPolygon: UserDefinedFunction = udf(_isMultiPolygon)
+
+  val isRoute: UserDefinedFunction = udf { (tags: Map[String, String]) =>
+    tags.contains("type") && tags("type") == "route"
+  }
+
+  private val MemberSchema = ArrayType(
+    StructType(
+      StructField("type", ByteType, nullable = false) ::
+        StructField("ref", LongType, nullable = false) ::
+        StructField("role", StringType, nullable = false) ::
+        Nil), containsNull = false)
+
+  private val _compressMemberTypes = (members: Seq[Row]) =>
+    members.map { row =>
+      val t = row.getAs[String]("type") match {
+        case "node" => NodeType
+        case "way" => WayType
+        case "relation" => RelationType
+      }
+      val ref = row.getAs[Long]("ref")
+      val role = row.getAs[String]("role")
+
+      Row(t, ref, role)
+    }
+
+  val compressMemberTypes: UserDefinedFunction = udf(_compressMemberTypes, MemberSchema)
+
+  class AssemblyException(msg: String) extends Exception(msg)
```

Comment:
@jbouffard from here down would benefit from some cleanup / refactoring / extraction.

Reply:
Okay, I'll take a look and see where I can improve things.

Reply:
Awesome. `ReversedCoordinateSequence`, `PartialCoordinateSequence`, and `VirtualCoordinateSequence` **definitely** shouldn't be public, since they don't provide immutable guarantees and implement some functions fast and loose.